js/source/geojson_source.js

```diff
@@ -37,56 +32,84 @@ module.exports = GeoJSONSource;
+ *
+ * // update
+ * var mySource = map.getSource('some id');
+ * mySource.setData({
+ *     data: {
+ *        "type": "FeatureCollection",
+ *        "features": [{
+ *            "type": "Feature",
+ *            "properties": { "name": "Null Island" },
+ *            "geometry": {
+ *                "type": "Point",
+ *                "coordinates": [ 0, 0 ]
+ *            }
+ *        }]
+ *     }
+ * })
+ *
  * map.removeSource('some id');  // remove
  */
-function GeoJSONSource(options) {
+function GeoJSONSource(id, options, dispatcher) {
     options = options || {};
+    this.id = id;
+    this.dispatcher = dispatcher;
 
     this._data = options.data;
 
     if (options.maxzoom !== undefined) this.maxzoom = options.maxzoom;
+    if (options.type) this.type = options.type;
 
     var scale = EXTENT / this.tileSize;
 
-    this.geojsonVtOptions = {
-        buffer: (options.buffer !== undefined ? options.buffer : 128) * scale,
-        tolerance: (options.tolerance !== undefined ? options.tolerance : 0.375) * scale,
-        extent: EXTENT,
-        maxZoom: this.maxzoom
-    };
-
-    this.cluster = options.cluster || false;
-    this.superclusterOptions = {
-        maxZoom: Math.min(options.clusterMaxZoom, this.maxzoom - 1) || (this.maxzoom - 1),
-        extent: EXTENT,
-        radius: (options.clusterRadius || 50) * scale,
-        log: false
-    };
-
-    this._pyramid = new TilePyramid({
-        tileSize: this.tileSize,
-        minzoom: this.minzoom,
-        maxzoom: this.maxzoom,
-        reparseOverscaled: true,
-        load: this._loadTile.bind(this),
-        abort: this._abortTile.bind(this),
-        unload: this._unloadTile.bind(this),
-        add: this._addTile.bind(this),
-        remove: this._removeTile.bind(this),
-        redoPlacement: this._redoTilePlacement.bind(this)
-    });
+    // sent to the worker, along with `url: ...` or `data: literal geojson`,
+    // so that it can load/parse/index the geojson data
+    // extending with `options.workerOptions` helps to make it easy for
+    // third-party sources to hack/reuse GeoJSONSource.
+    this.workerOptions = util.extend({
+        source: this.id,
+        cluster: options.cluster || false,
+        geojsonVtOptions: {
+            buffer: (options.buffer !== undefined ? options.buffer : 128) * scale,
+            tolerance: (options.tolerance !== undefined ? options.tolerance : 0.375) * scale,
+            extent: EXTENT,
+            maxZoom: this.maxzoom
+        },
+        superclusterOptions: {
+            maxZoom: Math.min(options.clusterMaxZoom, this.maxzoom - 1) || (this.maxzoom - 1),
+            extent: EXTENT,
+            radius: (options.clusterRadius || 50) * scale,
+            log: false
+        }
+    }, options.workerOptions);
+
+    this._sendDataToWorker(function done(err) {
+        if (err) {
+            this.fire('error', {error: err});
+            return;
+        }
+        this.fire('load');
+    }.bind(this));
```

Comment:
Good question.  In `master` it's being called in `update` (not `setData`), but now that you point it out, I don't see any reason not to put it in `setData`.  I think one implication of doing this would be that we would want to wait till the callback to fire `change` (which is currently fired synchronously in `setData`), but that seems fine.